<a href="https://colab.research.google.com/github/Lithikhaa/Dropouts/blob/main/Dropouts.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

In [ ]:
import numpy as np
import pandas as pd
import matplotlib.pyplot as plt
import seaborn as sns

In [ ]:
import warnings
warnings.filterwarnings('ignore')

In [ ]:
df =  pd.read_csv('/content/sonar_dataset.csv',header=None) #headernone is used to give a column name as 0 1 2 3 4 in the dataset
df

0       1       2       3       4       5       6       7       8   \
0    0.0200  0.0371  0.0428  0.0207  0.0954  0.0986  0.1539  0.1601  0.3109   
1    0.0453  0.0523  0.0843  0.0689  0.1183  0.2583  0.2156  0.3481  0.3337   
2    0.0262  0.0582  0.1099  0.1083  0.0974  0.2280  0.2431  0.3771  0.5598   
3    0.0100  0.0171  0.0623  0.0205  0.0205  0.0368  0.1098  0.1276  0.0598   
4    0.0762  0.0666  0.0481  0.0394  0.0590  0.0649  0.1209  0.2467  0.3564   
..      ...     ...     ...     ...     ...     ...     ...     ...     ...   
203  0.0187  0.0346  0.0168  0.0177  0.0393  0.1630  0.2028  0.1694  0.2328   
204  0.0323  0.0101  0.0298  0.0564  0.0760  0.0958  0.0990  0.1018  0.1030   
205  0.0522  0.0437  0.0180  0.0292  0.0351  0.1171  0.1257  0.1178  0.1258   
206  0.0303  0.0353  0.0490  0.0608  0.0167  0.1354  0.1465  0.1123  0.1945   
207  0.0260  0.0363  0.0136  0.0272  0.0214  0.0338  0.0655  0.1400  0.1843   

         9   ...      51      52      53      54      55      56      57  \
0    0.2111  ...  0.0027  0.0065  0.0159  0.0072  0.0167  0.0180  0.0084   
1    0.2872  ...  0.0084  0.0089  0.0048  0.0094  0.0191  0.0140  0.0049   
2    0.6194  ...  0.0232  0.0166  0.0095  0.0180  0.0244  0.0316  0.0164   
3    0.1264  ...  0.0121  0.0036  0.0150  0.0085  0.0073  0.0050  0.0044   
4    0.4459  ...  0.0031  0.0054  0.0105  0.0110  0.0015  0.0072  0.0048   
..      ...  ...     ...     ...     ...     ...     ...     ...     ...   
203  0.2684  ...  0.0116  0.0098  0.0199  0.0033  0.0101  0.0065  0.0115   
204  0.2154  ...  0.0061  0.0093  0.0135  0.0063  0.0063  0.0034  0.0032   
205  0.2529  ...  0.0160  0.0029  0.0051  0.0062  0.0089  0.0140  0.0138   
206  0.2354  ...  0.0086  0.0046  0.0126  0.0036  0.0035  0.0034  0.0079   
207  0.2354  ...  0.0146  0.0129  0.0047  0.0039  0.0061  0.0040  0.0036   

         58      59  60  
0    0.0090  0.0032   R  
1    0.0052  0.0044   R  
2    0.0095  0.0078   R  
3    0.0040  0.0117   R  
4    0.0107  0.0094   R  
..      ...     ...  ..  
203  0.0193  0.0157   M  
204  0.0062  0.0067   M  
205  0.0077  0.0031   M  
206  0.0036  0.0048   M  
207  0.0061  0.0115   M  

[208 rows x 61 columns]

In [ ]:
df.shape

(208, 61)

In [ ]:
df.isna().sum() #checking whether it has null values

0     0
1     0
2     0
3     0
4     0
     ..
56    0
57    0
58    0
59    0
60    0
Length: 61, dtype: int64

In [ ]:
df.columns #using header=none can generate the numerical value for column

Int64Index([ 0,  1,  2,  3,  4,  5,  6,  7,  8,  9, 10, 11, 12, 13, 14, 15, 16,
            17, 18, 19, 20, 21, 22, 23, 24, 25, 26, 27, 28, 29, 30, 31, 32, 33,
            34, 35, 36, 37, 38, 39, 40, 41, 42, 43, 44, 45, 46, 47, 48, 49, 50,
            51, 52, 53, 54, 55, 56, 57, 58, 59, 60],
           dtype='int64')

60 is the target variable from the column -- it contains the output column

In [ ]:
df[60].value_counts()#output row calculation

M    111
R     97
Name: 60, dtype: int64

In [ ]:
x = df.drop(60,axis='columns')
y = df[60]

In [ ]:
x

0       1       2       3       4       5       6       7       8   \
0    0.0200  0.0371  0.0428  0.0207  0.0954  0.0986  0.1539  0.1601  0.3109   
1    0.0453  0.0523  0.0843  0.0689  0.1183  0.2583  0.2156  0.3481  0.3337   
2    0.0262  0.0582  0.1099  0.1083  0.0974  0.2280  0.2431  0.3771  0.5598   
3    0.0100  0.0171  0.0623  0.0205  0.0205  0.0368  0.1098  0.1276  0.0598   
4    0.0762  0.0666  0.0481  0.0394  0.0590  0.0649  0.1209  0.2467  0.3564   
..      ...     ...     ...     ...     ...     ...     ...     ...     ...   
203  0.0187  0.0346  0.0168  0.0177  0.0393  0.1630  0.2028  0.1694  0.2328   
204  0.0323  0.0101  0.0298  0.0564  0.0760  0.0958  0.0990  0.1018  0.1030   
205  0.0522  0.0437  0.0180  0.0292  0.0351  0.1171  0.1257  0.1178  0.1258   
206  0.0303  0.0353  0.0490  0.0608  0.0167  0.1354  0.1465  0.1123  0.1945   
207  0.0260  0.0363  0.0136  0.0272  0.0214  0.0338  0.0655  0.1400  0.1843   

         9   ...      50      51      52      53      54      55      56  \
0    0.2111  ...  0.0232  0.0027  0.0065  0.0159  0.0072  0.0167  0.0180   
1    0.2872  ...  0.0125  0.0084  0.0089  0.0048  0.0094  0.0191  0.0140   
2    0.6194  ...  0.0033  0.0232  0.0166  0.0095  0.0180  0.0244  0.0316   
3    0.1264  ...  0.0241  0.0121  0.0036  0.0150  0.0085  0.0073  0.0050   
4    0.4459  ...  0.0156  0.0031  0.0054  0.0105  0.0110  0.0015  0.0072   
..      ...  ...     ...     ...     ...     ...     ...     ...     ...   
203  0.2684  ...  0.0203  0.0116  0.0098  0.0199  0.0033  0.0101  0.0065   
204  0.2154  ...  0.0051  0.0061  0.0093  0.0135  0.0063  0.0063  0.0034   
205  0.2529  ...  0.0155  0.0160  0.0029  0.0051  0.0062  0.0089  0.0140   
206  0.2354  ...  0.0042  0.0086  0.0046  0.0126  0.0036  0.0035  0.0034   
207  0.2354  ...  0.0181  0.0146  0.0129  0.0047  0.0039  0.0061  0.0040   

         57      58      59  
0    0.0084  0.0090  0.0032  
1    0.0049  0.0052  0.0044  
2    0.0164  0.0095  0.0078  
3    0.0044  0.0040  0.0117  
4    0.0048  0.0107  0.0094  
..      ...     ...     ...  
203  0.0115  0.0193  0.0157  
204  0.0032  0.0062  0.0067  
205  0.0138  0.0077  0.0031  
206  0.0079  0.0036  0.0048  
207  0.0036  0.0061  0.0115  

[208 rows x 60 columns]

In [ ]:
y

0      R
1      R
2      R
3      R
4      R
      ..
203    M
204    M
205    M
206    M
207    M
Name: 60, Length: 208, dtype: object

encoding the target column
R --> 1
M --> 0

In [ ]:
y = pd.get_dummies(y,drop_first=True)
y

R
0    1
1    1
2    1
3    1
4    1
..  ..
203  0
204  0
205  0
206  0
207  0

[208 rows x 1 columns]

In [ ]:
y.value_counts()

R
0    111
1     97
dtype: int64

In [ ]:
from sklearn.model_selection import train_test_split
X_train,X_test,y_train,y_test = train_test_split(x,y,test_size = 0.25,random_state = 1)

In [ ]:
df.shape

(208, 61)

In [ ]:
X_train.shape,y_train.shape,X_test.shape,y_test.shape

((156, 60), (156, 1), (52, 60), (52, 1))

In [ ]:
import tensorflow as tf
from tensorflow import keras


In [ ]:
model = keras.Sequential([
    keras.layers.Dense(60, input_dim=60, activation='relu'),
    keras.layers.Dense(30, activation='relu'), #30,15 for trail and error
    keras.layers.Dense(15, activation='relu'),
    keras.layers.Dense(1, activation='sigmoid')#binary classifer so 1
])
model.compile(loss='binary_crossentropy',optimizer ='adam', metrics=['accuracy'])#binary_crossentrophy this is used for calculate the loss --- it is binaryclassification so ..this binaryent
model.fit(X_train, y_train, epochs = 80, batch_size = 8)

Epoch 1/80
20/20 [==============================] - 1s 1ms/step - loss: 0.6785 - accuracy: 0.5577
Epoch 2/80
20/20 [==============================] - 0s 1ms/step - loss: 0.6569 - accuracy: 0.6026
Epoch 3/80
20/20 [==============================] - 0s 1ms/step - loss: 0.6332 - accuracy: 0.6731
Epoch 4/80
20/20 [==============================] - 0s 1ms/step - loss: 0.6103 - accuracy: 0.6731
Epoch 5/80
20/20 [==============================] - 0s 1ms/step - loss: 0.5809 - accuracy: 0.8013
Epoch 6/80
20/20 [==============================] - 0s 2ms/step - loss: 0.5588 - accuracy: 0.7372
Epoch 7/80
20/20 [==============================] - 0s 1ms/step - loss: 0.5190 - accuracy: 0.7756
Epoch 8/80
20/20 [==============================] - 0s 2ms/step - loss: 0.4800 - accuracy: 0.7500
Epoch 9/80
20/20 [==============================] - 0s 2ms/step - loss: 0.4765 - accuracy: 0.8013
Epoch 10/80
20/20 [==============================] - 0s 1ms/step - loss: 0.4254 - accuracy: 0.8205
Epoch 11/80
20/20 [

In [ ]:
model.evaluate(X_test,y_test)#76 percent of dataset only good balance 20 percent in dataset were not predicated well

2/2 [==============================] - 0s 11ms/step - loss: 0.8760 - accuracy: 0.7692


[0.876043975353241, 0.7692307829856873]

arr = np.array([[1, 2, 3], [4, 5, 6]])

newarr = arr.reshape(-1)
output -----
[1 2 3 4 5 6]  

(4,3) --- 4 row , 3 column as matrix in output

In [ ]:
y_pred = model.predict(X_test).reshape(-1)
print(y_pred[:10])

2/2 [==============================] - 0s 8ms/step
[5.1142276e-08 9.8023969e-01 9.9662304e-01 1.8194437e-04 9.9997282e-01
 9.9930054e-01 2.9955485e-01 9.9999833e-01 5.0041304e-05 9.9999148e-01]


In [ ]:
#for rounding the above predicated values
y_pred = np.round(y_pred)
print(y_pred[:10])


[0. 1. 1. 0. 1. 1. 0. 1. 0. 1.]


In [ ]:
#comparing the test and predicated value
y_test[:10]

R
186  0
155  0
165  0
200  0
58   1
34   1
151  0
18   1
202  0
62   1

In [ ]:
from sklearn.metrics import confusion_matrix , classification_report
print(classification_report(y_pred,y_test))

              precision    recall  f1-score   support

         0.0       0.85      0.74      0.79        31
         1.0       0.68      0.81      0.74        21

    accuracy                           0.77        52
   macro avg       0.77      0.78      0.77        52
weighted avg       0.78      0.77      0.77        52



keras dropout layer
it drops randomly

In [ ]:
modeld = keras.Sequential([
    keras.layers.Dense(60, input_dim=60, activation='relu'),
    keras.layers.Dropout(0.5),#it drops 50 percent of neurons
    keras.layers.Dense(30, activation='relu'), #30,15 for trail and error
    keras.layers.Dropout(0.5),
    keras.layers.Dense(15, activation='relu'),
    keras.layers.Dropout(0.5),
    keras.layers.Dense(1, activation='sigmoid')#binary classifer so 1

])
modeld.compile(loss='binary_crossentropy',optimizer ='adam', metrics=['accuracy'])#binary_crossentrophy this is used for calculate the loss --- it is binaryclassification so ..this binaryent
modeld.fit(X_train, y_train, epochs = 90, batch_size = 8)

Epoch 1/90
20/20 [==============================] - 1s 1ms/step - loss: 0.7409 - accuracy: 0.4679
Epoch 2/90
20/20 [==============================] - 0s 2ms/step - loss: 0.7046 - accuracy: 0.5192
Epoch 3/90
20/20 [==============================] - 0s 2ms/step - loss: 0.6779 - accuracy: 0.5256
Epoch 4/90
20/20 [==============================] - 0s 2ms/step - loss: 0.6804 - accuracy: 0.5641
Epoch 5/90
20/20 [==============================] - 0s 2ms/step - loss: 0.6978 - accuracy: 0.5577
Epoch 6/90
20/20 [==============================] - 0s 2ms/step - loss: 0.6947 - accuracy: 0.5385
Epoch 7/90
20/20 [==============================] - 0s 2ms/step - loss: 0.6775 - accuracy: 0.5385
Epoch 8/90
20/20 [==============================] - 0s 1ms/step - loss: 0.6839 - accuracy: 0.5064
Epoch 9/90
20/20 [==============================] - 0s 2ms/step - loss: 0.6548 - accuracy: 0.5962
Epoch 10/90
20/20 [==============================] - 0s 2ms/step - loss: 0.7284 - accuracy: 0.4744
Epoch 11/90
20/20 [

In [ ]:
modeld.evaluate(X_test,y_test)#accuracy was increased after dropout some neurons

2/2 [==============================] - 0s 8ms/step - loss: 0.3925 - accuracy: 0.7885


[0.39245864748954773, 0.7884615659713745]

In [ ]:
y_pred = modeld.predict(X_test).reshape(-1)
print(y_pred[:10])

2/2 [==============================] - 0s 4ms/step
[0.00148336 0.41874897 0.954364   0.04370234 0.99849224 0.9035444
 0.26411045 0.9977835  0.05715872 0.9992808 ]


In [ ]:
y_pred = np.round(y_pred)
print(y_pred[:10])

[0. 0. 1. 0. 1. 1. 0. 1. 0. 1.]


In [ ]:
from sklearn.metrics import confusion_matrix , classification_report
print(classification_report(y_pred,y_test))

              precision    recall  f1-score   support

         0.0       0.85      0.74      0.79        31
         1.0       0.68      0.81      0.74        21

    accuracy                           0.77        52
   macro avg       0.77      0.78      0.77        52
weighted avg       0.78      0.77      0.77        52



In [ ]:
y_pred = modeld.predict(X_test).reshape(-1)
y_pred = np.round(y_pred)


from sklearn.metrics import confusion_matrix , classification_report
print(classification_report(y_pred,y_test))

2/2 [==============================] - 0s 5ms/step
              precision    recall  f1-score   support

         0.0       0.89      0.75      0.81        32
         1.0       0.68      0.85      0.76        20

    accuracy                           0.79        52
   macro avg       0.78      0.80      0.78        52
weighted avg       0.81      0.79      0.79        52



in f1_score --- it is increased by -- 0.81 and 0.76 --by using dropouts